### PR TITLE
reset climbers using limit switches at the start

### PR DESCRIPTION
**Pull request by ojas-sanghi**

### CODE REVIEW

File: RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java [hunk 1]
```diff
@@ -329,6 +330,34 @@ public void periodic() {
...
     climberRightError.log(LogLevel.GENERAL, climberRight.getClosedLoopError());
 
+    boolean climbLeftRevLimitSwitchClosed = climberLeft.getSensorCollection().isRevLimitSwitchClosed();
+    boolean climbRightRevLimitSwitchClosed = climberRight.getSensorCollection().isRevLimitSwitchClosed();
+
+    if (!climbLeftRevLimitSwitchClosed)
+    {
+      if (!climbLeftEncoderZeroed)
+      {
+        climberLeft.set(ControlMode.PercentOutput, -0.1);
```
Comment (cal-miller-harvard) [inline RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java:340]: Don't hard code this motor value. Call it a constant like "CLIMBER_RETRACT_SLOW * CLIMBER_LEFT_MOTOR_INVERTED".

File: RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java [hunk 2]
```diff
@@ -329,6 +330,34 @@ public void periodic() {
...
+      }
+    }
+    else
+    {
+      climbLeftEncoderZeroed = true;
```
Comment (cal-miller-harvard) [inline RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java:345]: As far as I can tell, this just sets climbLeftEncoderZeroed, and doesn't actually zero the encoder. Does the encoder get zeroed somewhere else?

Also, don't you need to stop the motor when you're done zeroing?

**Review by cal-miller-harvard — CHANGES_REQUESTED** [review]
I left a couple minor comments, but I just want to point out that the code is very hard to follow as currently structured, since there's no explicit connection between the status of the system (e.g. idle, manual extend/ retract, unzeroed, zeroing, auto climb), the state of the inputs, and what the outputs are doing.

It may be unreasonable for you to rework the overall logic before your next event, but a fully specified finite state machine is a much better framework for this sort of code, to avoid errors like the ones I pointed out in my review comments.

In addition, it's not obvious what the system will do if you press e.g. the manual extend button while the system is zeroing; this is something you need to make sure won't cause undesirable behavior. A FSM model also makes those edge cases clear.

File: RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java (hunk 1 above)
Comment (ojas-sanghi) [reply RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java:340]: what's "CLIMBER_LEFT_MOTOR_INVERTED" mean in this example?

Comment (cal-miller-harvard) [reply RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java]: Whether a positive voltage extends or retracts the climber (so either -1 or 1)

File: RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java [hunk 3]
```diff
@@ -228,8 +233,6 @@ public void init() {
     // climberRight.setStatusFramePeriod(StatusFrameEnhanced.Status_10_Targets, 10);
 
     // zero sensors
-    //  TODO: actually use the limit switches to zero these at the start of the match! [[what does this mean]]
-    // You need to know the absolute encoder positions, relative to the limit switches. You can measure that by e.g. extend until you don’t see the lower limit, then retract until you see the lower limit, then set the encoder to zero
     climberLeft.getSensorCollection().setQuadraturePosition(0, MotorUtils.CONTROLLER_TIMEOUT_MS);
```
Comment (cal-miller-harvard) [inline RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java:236]: What does this line do?

File: RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java [hunk 4]
```diff
@@ -329,6 +332,38 @@ public void periodic() {
...
+    {
+      if (!climbLeftEncoderZeroed)
+      {
+        climberLeft.set(ControlMode.PercentOutput, climbRetractSlow.currentValue());
+      }
+    }
+    else
+    {
+      climberLeft.setSelectedSensorPosition(0);
+      climberLeft.set(0);
```
Comment (cal-miller-harvard) [inline RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java:348]: As far as I can tell, this sets the climber speed to zero whenever the limit switch is closed, which I don't think is your desired behavior. You can draw out a finite state machine state diagram for just the zeroing process, if you want to avoid logic bugs like this.

**Review by cal-miller-harvard — CHANGES_REQUESTED** [review]

File: RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java (hunk 3 above)
Comment (ojas-sanghi) [reply RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java:236]: ah turns it zeros the motors -- ill remove these

File: RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java (hunk 4 above)
Comment (ojas-sanghi) [reply RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java:348]: wait then what else would we want to do?

Comment (cal-miller-harvard) [reply RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java:344]: How can you ever extend the climber if the motor speed is set to zero whenever the retracted limit switch is pressed.

The correct behavior is setting the speed to zero once, at the same time as the isZeroed variable is set to false, then relying on the motor controller's configured limit switch behavior (not allowing + output when + switch is pressed and vise versa for - switch).

Comment (ojas-sanghi) [reply RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java]: > at the same time as the isZeroed variable is set to false

this is just at init, did you mean when it's set to true?

also won't the limit switch make the motor stop anyways?

**Review by cpostbitbuckets — APPROVED** [review]
